### PR TITLE
Resend SMS during login

### PR DESCRIPTION
**Pull request by vinaysshenoy**

### CODE REVIEW

File: app/src/test/java/org/simple/clinic/user/UserSessionTest.kt [hunk 1]
```diff
@@ -127,53 +132,110 @@ class UserSessionTest {
...
-    whenever(registrationApi.findUser("000")).thenReturn(Single.error(NullPointerException()))
+    try {
+      userSession.loginWithOtp("000000").blockingGet()
+    } catch (e: Throwable) {
+      thrownException = e
+    }
 
-    val result1 = userSession.findExistingUser("123").blockingGet()
-    assertThat(result1).isInstanceOf(FindUserResult.Found::class.java)
+    assertThat(thrownException).isNull()
```
Comment (saket) [inline app/src/test/java/org/simple/clinic/user/UserSessionTest.kt]: JUnit offers a slightly better way of expecting exceptions with `ExpectedException` rule. Do you want to use that? 

```kotlin
@Rule
val expectedException: ExpectedException = ExpectedException.none()

@Test
fun foo() {
  expectedException.expect(AnyException.class)
  expectedException.expectMessage("The expected message")
}
```

File: app/src/test/java/org/simple/clinic/user/UserSessionTest.kt [hunk 2]
```diff
@@ -127,53 +132,110 @@ class UserSessionTest {
...
+        listOf(Single.error<LoginResponse>(unauthorizedHttpError<LoginResponse>()), LoginResult.ServerError::class.java),
+        listOf(Single.error<LoginResponse>(IOException()), LoginResult.NetworkError::class.java)
+    )
+  }
 
-    val result3 = userSession.loginWithOtp("").blockingGet()
-    assertThat(result3).isInstanceOf(LoginResult.ServerError::class.java)
+  @Test
+  @Parameters(method = "parameters for triggering sync on login")
+  fun `sync should be triggered on successful login`(
```
Comment (saket) [inline app/src/test/java/org/simple/clinic/user/UserSessionTest.kt]: Should this be named as "sync should **only** be triggered on successful login"?

File: app/src/main/java/org/simple/clinic/enterotp/EnterOtpScreenController.kt [hunk 3]
```diff
@@ -97,4 +98,33 @@ class EnterOtpScreenController @Inject constructor(private val userSession: User
         .compose(NewlyVerifiedUser())
         .map { { ui: Ui -> ui.goBack() } }
   }
+
+  private fun resendSms(events: Observable<UiEvent>): Observable<UiChange> {
+
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/enterotp/EnterOtpScreenController.kt]: Empty line. Static analysis should catch this :(

File: app/src/main/java/org/simple/clinic/enterotp/EnterOtpScreenController.kt [hunk 4]
```diff
@@ -97,4 +98,33 @@ class EnterOtpScreenController @Inject constructor(private val userSession: User
...
+                Observable.merge(handleRequestOtpResult(loginResult), hideProgressOnLoginResult(loginResult))
+              }
+              .startWith { ui: Ui ->
+                ui.hideError()
+                ui.showProgress()
+              }
+        }
+  }
+
+  private fun handleRequestOtpResult(loginResult: LoginResult): Observable<UiChange> {
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/enterotp/EnterOtpScreenController.kt]: Please don't use `handle` for naming functions. All functions are anyway handling something. :)

**Comment by saket** [comment]
@vinaysshenoy, the spacing between `RESEND SMS` button and the card's bottom doesn't match Zeplin. Can you check this?

File: app/src/test/java/org/simple/clinic/user/UserSessionTest.kt [hunk 5]
```diff
@@ -90,6 +92,9 @@ class UserSessionTest {
 
   @Before
   fun setUp() {
+    // Used for overriding IO scheduler for sync call on login.
+    RxJavaPlugins.setIoSchedulerHandler { Schedulers.trampoline() }
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/user/UserSessionTest.kt]: Done.

File: app/src/test/java/org/simple/clinic/user/UserSessionTest.kt (hunk 1 above)
Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/user/UserSessionTest.kt]: Oh wait, this test is incorrect. I need to check the result of the `loginWithOtp` call and it should be `Success` in all cases. Thanks for calling attention to this!

Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/user/UserSessionTest.kt]: Fixed.

File: app/src/main/java/org/simple/clinic/enterotp/EnterOtpScreenController.kt (hunk 3 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/enterotp/EnterOtpScreenController.kt]: Fixed.

File: app/src/main/java/org/simple/clinic/enterotp/EnterOtpScreenController.kt (hunk 4 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/enterotp/EnterOtpScreenController.kt]: Renamed to `showMessageOnLoginResult`.

**Comment by vinaysshenoy** [comment]
@sanchita Done.

**Pull request merged**